finality-aleph/src/data_io.rs

```diff
@@ -359,82 +1001,228 @@ where
+            let finalized_block =
+                BlockHashNum::<B>::new(client_info.finalized_hash, client_info.finalized_number);
+
+            if let Some(prev) = &self.prev_chain_info {
+                if prev.best_block_in_session == *best_block
+                    && prev.highest_finalized == finalized_block
+                {
+                    // This is exactly the same state that we processed last time in update_data.
+                    // No point in recomputing.
+                    return;
                 }
-                prev_best_hash = new_best_header.hash();
-                prev_best_number = *new_best_header.number();
             }
-            _ = &mut exit => {
-                debug!(target: "afa", "Task for refreshing best chain received exit signal. Terminating.");
+            // Update the info for the next call of update data.
+            self.prev_chain_info = Some(ChainInfoForData {
+                best_block_in_session: best_block.clone(),
+                highest_finalized: finalized_block.clone(),
+            });
+            let mut curr_block = best_block.clone();
+
+            if curr_block.num == finalized_block.num {
+                // We don't have anything to propose, we go ahead with an empty propasal.
+                *self.data_to_propose.lock() = AlephData::HeadProposal(AlephProposal::<B>::new(
+                    Vec::new(),
+                    self.min_block_num,
+                ));
+                return;
+            }
+            if curr_block.num < finalized_block.num {
+                // Because of the client synchronization, in extremely rare cases this could happen.
+                warn!(target: "afa", "Error updating data. curr block {:?}, finalized {:?}", curr_block, finalized_block);
                 return;
             }
+            let mut branch: Vec<B::Hash> = Vec::new();
+            while curr_block.num > finalized_block.num {
+                if curr_block.num - finalized_block.num
+                    <= <NumberFor<B>>::saturated_from(MAX_DATA_BRANCH_LEN)
+                {
+                    branch.push(curr_block.hash);
+                }
+                curr_block = get_parent(self.client.clone(), &curr_block)
+                    .expect("block of num >= 1 must have a parent")
+            }
+            if curr_block.hash == finalized_block.hash {
+                let num_last = finalized_block.num + <NumberFor<B>>::saturated_from(branch.len());
+                // The hashes in `branch` are ordered from top to bottom -- need to reverse.
+                branch.reverse();
+                *self.data_to_propose.lock() =
+                    AlephData::HeadProposal(AlephProposal::<B>::new(branch, num_last));
+            } else {
+                // By backtracking from the best block we reached a block conflicting with best finalized.
+                // This is most likely a bug, or some extremely unlikely synchronization issue of the client.
```

Comment:
This is kind of like that, but this component is responsible for creating our own proposal, and they definitely should not be hopeless at this stage :) 

This concrete situation is something like that:
1. We ask select_chain for the best_block.
2. We reduce it so that it doesn't stick above the last_block of this session.
3. Then we start traversing parents and realize that we don't `hit` the last finalized block! 

There are no delays between 1, 2, 3, so this should not happen, unless there was some extremely unlucky reorg happening in the background.